### DISCUSSION TITLE
Plot well logs with pyvista + matplotlib

### DISCUSSION THREAD

**Discussion by jc-barreto** [pyvista/pyvista#2094] [Q&A]
Is there a way to introduce a topography created with pyvista into a 3D scatter plot created with matplotlib?

I need to visualize some borehole, and after trying for a while with another library, I give up.  I would like to introduce this topography (a .tif file):
![image](https://user-images.githubusercontent.com/42352209/151538455-cdea6934-094a-48f8-ab19-8616c493c320.png)
into my 3D scatter plot:
![image](https://user-images.githubusercontent.com/42352209/151538660-a89d6715-e37a-429e-b6aa-27e635fae005.png)
my borehole file is in the format:  x, y, z, label

**Comment by banesullivan** [comment]
We can help you go the other way (plotting your boreholes in PyVista with the topography surface) 😉 

Matplotlib is not going to handle your 3D topography surface well, and frankly I wouldn't recommend trying. If you're inclined, I'd have to point your to MPL's docs: 

- https://matplotlib.org/stable/gallery/mplot3d/trisurf3d_2.html
- https://matplotlib.org/stable/tutorials/toolkits/mplot3d.html#mpl_toolkits.mplot3d.Axes3D.plot_trisurf

Accessing the data from PyVista should be straightforward. Vertices come from `mesh.points` and triangles can be fetched for `PolyData` by:

```py
triangles = mesh.faces.reshape((-1,4))[:, 1:]  # Assumes triangle only pyvista.PolyData
```

**Comment by banesullivan** [comment]
> my borehole file is in the format: x, y, z, label

You can easily add these to your PyVista plot of your terrain surface. 

Load your x, y, z, label data into numpy arrays, then:

```py
boreholes = pv.PolyData(np.c_[x,y,z])
boreholes['labels'] = labels
```

and then add them to your existing plotter with the topography using `.add_mesh(boreholes)`

**Reply by banesullivan** [reply]
FYI:

- https://github.com/pyvista/pyvista-support/issues/25
- https://github.com/pyvista/pyvista-support/issues/303

**Reply by jc-barreto** [reply]
I did what you suggested, but I didn't get the boreholes, no error, but in the image doens't show up the boreholes:
this is my code:

```
elevArray = data.loc[:,['X','Y','Z']].to_numpy()
boreholes = pv.PolyData(elevArray)
labels = data.loc[:,['labels']].to_numpy()

sargs = dict(fmt="%.0f", color='black')


p = pv.Plotter(notebook=True)
p.add_mesh(mesh=topo, cmap='gist_earth', scalar_bar_args=sargs, clim=[1220, 1350])
p.add_mesh(boreholes)
p.set_background('white')
p.show_grid(color='black')
p.set_scale(1,1,0.1)
p.show()
```

**Reply by kusuma2022** [reply]
@jc-barreto : thanks for bringing this discussion and example of add_mesh. I manage to plot my data . I use  "np.c_" instead of ".to numpy"
@banesullivan : thanks for the explanation and the references, i was looking for this for quite sometimes and finally I found here

**Reply by jc-barreto** [reply]
@kusuma2022 would you mind putting your code here? 
I didn't manage to plot my borehole yet...

**Reply by kusuma2022** [reply]
here is my code:
boreholdexyz = df[['x_index','y_index', 'z_index']] 
boreholefacies = df[['Facies']]

boreholes = pv.PolyData(np.c_[boreholdexyz])
boreholes['labels'] = boreholefacies
p.add_mesh(boreholes,point_size=10)
